### DISCUSSION TITLE
Improve DB2-Platform support for Ebean

### DISCUSSION THREAD

**Discussion by rPraml** [ebean-orm/ebean#2484] [General]
My company plans to use DB2 with ebean, but the current DB2 support of Ebean is not good enough. Some features are missing, or are not working properly.
DB2 platform exists from the beginning (at least since 2013), but there were no automatic tests in the past. So all new features of ebean are untested on the DB2 platform.

Now, with 2021/2022 we are able to spawn a DB2-docker container and run over 3000 tests against this platform.
After fixing the issues with constraint names, ~95% of them were passing. So we investigated some time to fix the rest of them.

**Who is using DB2?**

There are different DB2 families by IBM, that are often mixed up: https://en.wikipedia.org/wiki/IBM_Db2_Family
- DB2 Database (also known as DB2 LUW). Current version 11.5. (We use the free [DB2 community edition 11.5.6.0a](https://hub.docker.com/r/ibmcom/db2) to run the tests)
- DB2 for z/OS, Current version 12.1 
- DB2 for i (formerly AS/400), Current version 7.4
and they also have different limitations, features and sometimes will reqire also different SQL statements.

So I think, we should make a platform for each of these "dialects". @bigjimlefou has already started to start with a [DB2ForIPlatform](https://github.com/bigjimlefou/ebean/blob/master/ebean-api/src/main/java/io/ebean/config/dbplatform/db2/DB2ForIPlatform.java)

It would be also great, if some people can help to run the tests against iSeries or z/OS (I have no access to these platforms)

**List of fixes we have made**

- [Update table & constraint name length to 128](https://github.com/FOCONIS/ebean/commit/1ff0d781707c8e9edb9440ed46e6020e740203a2) 
We had to update the max column length to 128, to get the tests working. Without this change, existing ebean tests will not start.
This will be a **breaking change** for users that are already using DB2 (should we force these users to Db2LegacyPlatform or should we force them to migrate their schema?)
@bigjimlefou ran in the same issue here [DB2ForIPlatform](https://github.com/bigjimlefou/ebean/blob/master/ebean-api/src/main/java/io/ebean/config/dbplatform/db2/DB2ForIPlatform.java)
- [FIX Concat operator](https://github.com/FOCONIS/ebean/commit/1f6916775bb24c7492cc00781539de0ac127f994) we have to use here the same operator as in Oracle.
- [Fix Sequence generator](https://github.com/FOCONIS/ebean/commit/f7be1de6df4a59ec27cc6950bb6517006c9e04bd) the SQL was fixed to work with DB2 LUW (did it ever work when multiple ids were requested?)
- [Fix insert default values](https://github.com/FOCONIS/ebean/commit/cc21ffcca9c4479321de7db6262eb2d22c995ed5) 
  See [Insert statement](https://www.ibm.com/docs/en/db2/11.5?topic=statements-insert).
  This should be compatible with [IBMi](https://www.ibm.com/docs/en/i/7.1?topic=language-inserting-rows-using-insert-statement) and [z/OS](https://www.ibm.com/docs/en/db2-for-zos/11?topic=statements-insert)
- [Fix getGenerateKeys](https://github.com/FOCONIS/ebean/commit/29eb07e78f4a180e331e61975d6d56220bfd1dbe)
  **This is a really ugly workaround with reflection** With DB2 we require a workaround to get the IDs in a batch insert. See [StackOverflow](https://stackoverflow.com/questions/41725492/how-to-get-auto-generated-keys-of-batch-insert-statement)
  @rbygrave, what do you think of this change? Should we route that to the Db2-Platform
- [Fix raw like](https://github.com/FOCONIS/ebean/commit/47d5405b7ce33566309511f06c3be25ad8e8209d) nothing critical
- [Enable persistBatchOnCascade](https://github.com/FOCONIS/ebean/commit/5a59298cb8056aefe9ce2afc8022f5ff8fa82844) we enabled the persistBatchOnCascade again, as it works now.
- [SqlCodeTranslator must check exception chain](https://github.com/FOCONIS/ebean/commit/442732d70da40261ae22b4d5eb5ce726e99f069a) see comment in code
- [Fix Db2SqlLimiter](https://github.com/FOCONIS/ebean/commit/193617a107da2d98e4ba03c4140e28ec23bb4034) The limiter did not support "offset". It uses the same syntax as oracle now.
- [Support for unique indices with null columns](https://github.com/FOCONIS/ebean/commit/23a5d577a6aafe015ffaca74bda6342b8ecd1829)
  `exclude null keys` is supported since [DB2 11.1 LUW](https://www.ibm.com/docs/en/db2/11.1?topic=statements-create-index#sdx-synid_exclude_null_keys), 
  [DB2 11 for z/OS](https://www.ibm.com/docs/en/db2-for-zos/11?topic=statements-create-index) - unfortunately, for [IBMi](https://www.ibm.com/docs/en/i/7.1?topic=statements-create-index) we need a different syntax with `UNIQUE WHERE NOT NULL`
  *It may be not available for elder versions*
- [truncate needs commit after each statement](https://github.com/FOCONIS/ebean/commit/ea9481477cdf04eaf00b7aba5d57b27ce2324729)
  @rbygrave as far as I know, truncate should be used only in tests. Do you think, we can do a commit after each truncate (and do not check the platform)
- [Experimental: Add QueryPlanLogger for DB2](https://github.com/FOCONIS/ebean/commit/ae79dd92ec5a1fe706223307fa5a5edf7df902ff) We add a queryPlanLogger for DB2, this will require also some [extraDdl](https://github.com/FOCONIS/ebean/commit/0d8d255f178390e2e3d8212aad2f48e11611b849)
- [DB2 does not support ids in compressed form](https://github.com/FOCONIS/ebean/commit/2dc9087eb32bff59a55e49a040fe0d08aa4f327e)
- [Do not try to update deferred null values](https://github.com/FOCONIS/ebean/commit/5f707053d362fd7bd010fc63b01c8a0c49ffdd3f) We found some bugs discovered by TestPrinterStateIssue, where ebean tries to update null values. @rbygrave maybe you can take a look at this?

This should be the complete list of fixes, now 3333 tests are passing in our CI
There are some more changes in test classes (asserts against generated sql etc) you may see in this commit list https://github.com/FOCONIS/ebean/commits/0d8d255f178390e2e3d8212aad2f48e11611b849

**What to do next?**
- set up a github action for DB2 LUW
- get some feedback from ebean & DB2 users about their versions and platforms
- decide, if we can change the max column length to 128 (and introduce a legacy platform) or if we fix the tests and use shorter column names etc, so that we don't get collisions in indices. (may depend on community feedback)
- cherry pick the above commits and create separate issues/PRs
- tests for other platform (z/OS, IBMi)
- review skipped tests (there are 126 now), if they might be enabled for DB2
- add "DB2 compatible" badge to Ebean :)

cheers
Roland

**Comment by rbygrave** [comment]
> So I think, we should make a platform for each of these "dialects". @bigjimlefou has already started to start with a DB2ForIPlatform

That might be the first step. Create the 3 platforms and perhaps a 4th called "legacy".  Then I can release that artifact such that PR's can depend on it.
